libraries/common/test.effekt

```diff
@@ -80,6 +323,185 @@ def test(name: String) { body: => Unit / Assertion } = {
+        do stop()
+      }
+  }
+}
+
+def forall[A, B](name: String, n: Int)
+  { body: (A, B) => Unit / Assertion }: Unit / {Test, Generator[A], Generator[B]} = {
+
+  val startTime = bench::relativeTimestamp()
+  var successCounter = 0
+  with boundary;
+  with val x = for[A] {with limit[A](n + 1); do generate[A]()}
+  with val y = for[B] {with limit[B](n + 1); do generate[B]()}
+
+  try {
+    body(x, y)
+    if(successCounter == n) {
+      val duration = Duration::diff(startTime, bench::relativeTimestamp())
+      do successForall(name, n, duration) 
+      do stop()
+    }
+  } with Assertion { 
+    def assert(condition, msg) = 
+      if(condition) {
+        successCounter = successCounter + 1 
+        resume(())
+      }
+      else {
+        val duration = Duration::diff(startTime, bench::relativeTimestamp())
+        do failureForall(name, successCounter, msg ++ "\n  failed on inputs:\n    1. " ++ genericShow(x) ++ "\n    2. " ++ genericShow(y), duration)
+        do stop()
+      }
+  }
+}
+
+def forall[A, B, C](name: String, n: Int)
+  { body: (A, B, C) => Unit / Assertion }: Unit / {Test, Generator[A], Generator[B], Generator[C]} = {
+
+  val startTime = bench::relativeTimestamp()
+  var successCounter = 0
+  with boundary;
+  with val x = for[A] {with limit[A](n + 1); do generate[A]()}
+  with val y = for[B] {with limit[B](n + 1); do generate[B]()}
+  with val z = for[C] {with limit[C](n + 1); do generate[C]()}
+
+  try {
+    body(x, y, z)
+    if(successCounter == n) {
+      val duration = Duration::diff(startTime, bench::relativeTimestamp())
+      do successForall(name, n, duration) 
+      do stop() 
+      }
+  } with Assertion { 
+    def assert(condition, msg) = 
+      if(condition) {
+        successCounter = successCounter + 1 
+        resume(())
+      }
+      else {
+        val duration = Duration::diff(startTime, bench::relativeTimestamp())
+        do failureForall(name, successCounter, msg ++ "\n  failed on inputs:\n    1. " ++ genericShow(x) ++ "\n    2. " ++ genericShow(y) ++ "\n    3. " ++ genericShow(z), duration)
+        do stop()
+      }
+  }
+}
+
+/// exists quantifier for property-based testing
+def exists[A](name: String, n: Int) 
+  { body: A => Unit / Assertion }: Unit / {Test, Generator[A]} = {
+
+  val startTime = bench::relativeTimestamp()
+  var triedCounter = 0
+  with boundary
+  with val x = for[A] {with limit[A](n + 1); do generate[A]()}
+
+  try {
+    body(x)
+    if(triedCounter == n) {
+      val duration = Duration::diff(startTime, bench::relativeTimestamp())
+      do failureExists(name, triedCounter, duration) 
+    }
+  } with Assertion { 
+    def assert(condition, msg) = 
+      if (condition) {
+        val duration = Duration::diff(startTime, bench::relativeTimestamp())
+        do successExists(name, triedCounter, " example value:\n    1. " ++ genericShow(x), duration)
```

Comment:
Same as in `forall`, in general, this needs to take a `show` parameter which is then applied on the tested value.